Sources/BSON/BSON.swift

```diff
@@ -77,6 +77,40 @@ public enum BSON {
         }
     }
 
+    /// Initialize a `BSON` from ExtendedJSON
+    /// Parameters:
+    ///   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for any `BSONValue`.
+    ///   - `keyPath`: an array of `Strings`s containing the enclosing JSON keys of the current json being passed in.
+    ///              This is used for error messages.
+    ///
+    /// Throws:
+    ///   - `DecodingError` if `json` is malformatted extended json
+    internal init(fromExtJSON json: JSON, keyPath: [String]) throws {
+        // Spec requires that we try int32, try int64, then try double for decoding numbers
+        if let int32 = try Int32(fromExtJSON: json, keyPath: keyPath) {
+            self = int32.bson
+            return
+        }
+        if let int64 = try Int64(fromExtJSON: json, keyPath: keyPath) {
+            self = int64.bson
+            return
+        }
+        for bsonType in BSON.allBSONTypes.values {
+            if bsonType == Int32.self || bsonType == Int64.self || bsonType == BSONDocument.self {
```

Comment:
I was having some issues with the way you suggested because the `BSONValue`s don't conform to Equatable. Patrick and I discussed this and decided to leave it with the guard

Reply:
sgtm